src/lib.rs

```diff
@@ -108,34 +105,36 @@ pub extern "C" fn release_cache(cache: *mut cache_t) {
 pub extern "C" fn create(cache: *mut cache_t, wasm: Buffer, err: Option<&mut Buffer>) -> Buffer {
     let r = match to_cache(cache) {
         Some(c) => catch_unwind(AssertUnwindSafe(move || do_create(c, wasm)))
-            .unwrap_or_else(|_| Panic {}.fail()),
-        None => EmptyArg { name: CACHE_ARG }.fail(),
+            .unwrap_or_else(|_| Err(Error::make_panic())),
+        None => Err(Error::make_empty_arg(CACHE_ARG)),
     };
     let check = handle_c_error(r, err);
     Buffer::from_vec(check.to_vec())
 }
 
 fn do_create(cache: &mut CosmCache<DB, GoApi, GoQuerier>, wasm: Buffer) -> Result<Checksum, Error> {
-    let wasm = unsafe { wasm.read() }.ok_or_else(|| empty_err(WASM_ARG))?;
-    cache.save_wasm(wasm).context(WasmErr {})
+    let wasm = unsafe { wasm.read() }.ok_or_else(|| Error::make_empty_arg(WASM_ARG))?;
+    let checksum = cache.save_wasm(wasm)?;
+    Ok(checksum)
```

Comment:
you can replace this with one of these:
```rust
Ok(cache.save_wasm(wasm)?)
cache.save_wasm(wasm).map_err(Into::into)
```

Reply:
Yeah I know, but wanted to avoid too much complexity in each line

Reply:
I remember some of your code where you used a comment in such a case. I agree it is not trivial to understand the flow. So instead of a comment I used a variable, which explains the intermediate state.